content/blog/0045-from-react-to-effect.mdx

```diff
@@ -0,0 +1,422 @@
+As we can see now we have our "Hello World" message printed, just in the same way as React we can compose different Effects together into more complex programs, to do that we will use a generator function:
+
+```ts
+import { Effect } from "effect"
+
+const print = (message: string) =>
+  Effect.sync(() => {
+    console.log(message)
+  })
+
+const printMessages = Effect.gen(function*() {
+  yield* print("Hello World")
+  yield* print("We're getting messages")
+})
+
+Effect.runPromise(printMessages)
+```
+
+<br />
+<a href="https://effect.website/play#b5058f3ac17b" target="_blank">Open in Playground</a>
+
+You can mentally map `yield*` to `await` and `Effect.gen(function*() { })` to `async function() {}` with the sole difference that in order to pass arguments you'd need to define a new lambda, for example:
+
+```ts
+import { Effect } from "effect"
+
+const print = (message: string) =>
+  Effect.sync(() => {
+    console.log(message)
+  })
+
+const printMessages = (messages: number) => Effect.gen(function*() {
+  for (let i = 0; i < messages; i++) {
+    yield* print(`message: ${i}`)
+  }
+})
+
+Effect.runPromise(printMessages(10))
+```
+
+<br />
+<a href="https://effect.website/play#3c5c45b473a4" target="_blank">Open in Playground</a>
+
+Similarly to how React components compose together we've now seen that also Effects can easily be composed, in the same way as we were able to raise an error from a component we can just as easily raise an error from an Effect and recover it from a parent Effect:
+
+```ts
+import { Effect } from "effect"
+
+const print = (message: string) =>
+  Effect.sync(() => {
+    console.log(message)
+  })
+
+class InvalidRandom extends Error {
+  message = "Invalid Random Number"
+}
+
+const printOrFail = Effect.gen(function*() {
+  if (Math.random() > 0.5) {
+    yield* print("Hello World")
+  } else {
+    yield* Effect.fail(new InvalidRandom())
+  }
+})
+
+const program = printOrFail.pipe(
+  Effect.catchAll((e) => print(`Error: ${e.message}`)),
+  Effect.repeatN(10)
+)
+
+Effect.runPromise(program)
+```
+
+<br />
+<a href="https://effect.website/play#a35378f5d2ab" target="_blank">Open in Playground</a>
+
+The above code randomly fails with an error that we recover from a parent Effect, in this case the recovery logic is to print the error message, differently from `React` we can see here that errors are type safe, in fact in our `catchAll` we know that `e` is of type `InvalidRandom`, this is possible because Effect uses inference to understand which error cases your program may encounter and represent those cases in its type, you can check the type of `printOrFail` and you'll see: `Effect<void, InvalidRandom, never>` which means that this Effect will return `void` if successful but may fail for `InvalidRandom`. You can compose Effects that can fail for different reasons and your final Effect will list all possible errors in a union, so you'd get for example `Effect<number, InvalidRandom | NetworkError | ..., never>`.
```

Comment:
```suggestion
The above code will randomly fail with a `InvalidRandom` error, which we then recover from a parent Effect using `Effect.catchAll`. In this case, the recovery logic is to simply log the error message to the console.

However, what separates Effect from `React` is that errors are 100% type safe - within our `Effect.catchAll`, we know that `e` is of type `InvalidRandom`. This is possible because Effect uses type inference to understand which error cases your program may encounter and represent those cases in its type. If you check the type of `printOrFail`, you will see: 

`Effect<void, InvalidRandom, never>` 

which means that this Effect will return `void` if successful but may also fail with an `InvalidRandom` error. 

When you compose Effects that may fail for different reasons, your final Effect will list all possible errors in a union, so you would see something like the following in the type: 

`Effect<number, InvalidRandom | NetworkError | ..., never>`.
```